// react-noop-renderer/index.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('react'), require('scheduler')) :
	typeof define === 'function' && define.amd ? define(['react', 'scheduler'], factory) :
	(global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.ReactNoopRenderer = factory(global.React, global.Scheduler));
})(this, (function (React, Scheduler) { 'use strict';

	function _interopNamespaceDefault(e) {
		var n = Object.create(null);
		if (e) {
			Object.keys(e).forEach(function (k) {
				if (k !== 'default') {
					var d = Object.getOwnPropertyDescriptor(e, k);
					Object.defineProperty(n, k, d.get ? d : {
						enumerable: true,
						get: function () { return e[k]; }
					});
				}
			});
		}
		n.default = e;
		return Object.freeze(n);
	}

	var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);
	var Scheduler__namespace = /*#__PURE__*/_interopNamespaceDefault(Scheduler);

	const FunctionComponent = 0;
	const HostRoot = 3;
	const HostComponent = 5;
	const HostText = 6;
	const Fragment = 7;

	const NoFlags = 0b0000000;
	const Placement = 0b0000001;
	const Update = 0b0000010;
	const ChildDeletion = 0b0000100;
	const PassiveEffect = 0b0001000;
	const MutationMask = Placement | Update | ChildDeletion;
	const PassiveMask = PassiveEffect | ChildDeletion;

	const SyncLane = 0b0001;
	const NoLane = 0b0000;
	const NoLanes = 0b0000;
	function mergeLanes(laneA, laneB) {
	    return laneA | laneB;
	}
	function requestUpdateLane() {
	    return SyncLane;
	}
	function getHighestPriorityLane(lanes) {
	    return lanes & -lanes;
	}
	function markRootFinished(root, lane) {
	    root.pendingLanes &= ~lane;
	}

	class FiberNode {
	    type;
	    tag;
	    pendingProps;
	    key;
	    stateNode;
	    return;
	    sibling;
	    child;
	    index;
	    ref;
	    memorizedProps;
	    memorizedState;
	    alternate;
	    flags;
	    subTreeFlags;
	    updateQueue;
	    deletions;
	    constructor(tag, pendingProps, key) {
	        this.tag = tag;
	        this.key = key || null;
	        this.stateNode = null;
	        this.type = null;
	        this.return = null;
	        this.sibling = null;
	        this.child = null;
	        this.index = 0;
	        this.ref = null;
	        this.pendingProps = pendingProps;
	        this.memorizedProps = null;
	        this.flags = NoFlags;
	        this.subTreeFlags = NoFlags;
	        this.updateQueue = null;
	        this.memorizedState = null;
	        this.alternate = null;
	        this.deletions = null;
	    }
	}
	class FiberRootNode {
	    container;
	    current;
	    finishedWork;
	    pendingLanes;
	    finishedLane;
	    pendingPassiveEffects;
	    constructor(container, hostRootFiber) {
	        this.current = hostRootFiber;
	        this.container = container;
	        hostRootFiber.stateNode = this;
	        this.finishedWork = null;
	        this.pendingLanes = NoLanes;
	        this.finishedLane = NoLane;
	        this.pendingPassiveEffects = {
	            unmount: [],
	            update: []
	        };
	    }
	}
	const createWorkingProgress = (current, pendingProps) => {
	    let wip = current.alternate;
	    if (wip === null) {
	        wip = new FiberNode(current.tag, pendingProps, current.key);
	        wip.stateNode = current.stateNode;
	        wip.alternate = current;
	        current.alternate = wip;
	    }
	    else {
	        wip.deletions = null;
	        wip.flags = NoFlags;
	        wip.subTreeFlags = NoFlags;
	        wip.pendingProps = pendingProps;
	    }
	    wip.type = current.type;
	    wip.updateQueue = current.updateQueue;
	    wip.child = current.child;
	    wip.memorizedProps = current.memorizedProps;
	    wip.memorizedState = current.memorizedState;
	    return wip;
	};
	function createFiberFromElement(element) {
	    const { type, key, props } = element;
	    let fiberTag = FunctionComponent;
	    if (typeof type === 'string') {
	        fiberTag = HostComponent;
	    }
	    else if (typeof type !== 'function' && true) {
	        console.warn('未定义的type类型', element);
	    }
	    const fiber = new FiberNode(fiberTag, props, key);
	    fiber.type = type;
	    return fiber;
	}
	function createFiberFromFragment(element, key) {
	    const fiber = new FiberNode(Fragment, element, key);
	    return fiber;
	}

	function createUpdate(action, lane) {
	    return {
	        action,
	        lane,
	        next: null
	    };
	}
	function createUpdateQueue() {
	    return {
	        shared: {
	            pending: null
	        },
	        dispatch: null
	    };
	}
	function enQueueUpdate(updateQueue, update) {
	    const pending = updateQueue.shared.pending;
	    if (pending === null) {
	        update.next = update;
	    }
	    else {
	        update.next = pending.next;
	        pending.next = update;
	    }
	    updateQueue.shared.pending = update;
	}
	function processUpdateQueue(baseState, pendingUpdate, renderLane) {
	    const result = {
	        memorizedState: baseState
	    };
	    if (pendingUpdate !== null) {
	        const first = pendingUpdate.next;
	        let pending = pendingUpdate.next;
	        do {
	            const updateLane = pending.lane;
	            if (updateLane === renderLane) {
	                const action = pendingUpdate.action;
	                if (action instanceof Function) {
	                    baseState = action(baseState);
	                }
	                else {
	                    baseState = action;
	                }
	            }
	            else {
	                {
	                    console.error('不应该进入updateLane !== renderLane这个逻辑');
	                }
	            }
	            pending = pending?.next;
	        } while (pending !== first);
	    }
	    result.memorizedState = baseState;
	    return result;
	}

	let instanceCounter = 0;
	// export const createInstance = (type: string, props: any): Instance => {
	const createInstance = (type, props) => {
	    const instance = {
	        id: instanceCounter++,
	        type,
	        children: [],
	        parent: -1,
	        props
	    };
	    return instance;
	};
	const appendInitialChild = (parent, child) => {
	    // id
	    const prevParentID = child.parent;
	    const parentID = 'rootID' in parent ? parent.rootID : parent.id;
	    if (prevParentID !== -1 && prevParentID !== parentID) {
	        throw new Error('不能重复挂载child');
	    }
	    child.parent = parentID;
	    parent.children.push(child);
	};
	const createTextInstance = (content) => {
	    const instance = {
	        text: content,
	        id: instanceCounter++,
	        parent: -1
	    };
	    return instance;
	};
	const appendChildToContainer = (parent, child) => {
	    // id
	    const prevParentID = child.parent;
	    if (prevParentID !== -1 && prevParentID !== parent.rootID) {
	        throw new Error('不能重复挂载child');
	    }
	    child.parent = parent.rootID;
	    parent.children.push(child);
	};
	function commitUpdate(fiber) {
	    switch (fiber.tag) {
	        case HostText:
	            // eslint-disable-next-line no-case-declarations
	            const text = fiber.memorizedProps.content;
	            return commitTextUpdate(fiber.stateNode, text);
	        default:
	            {
	                console.warn('未实现的Update类型', fiber);
	            }
	            break;
	    }
	}
	function commitTextUpdate(textInstance, content) {
	    textInstance.text = content;
	}
	function removeChild(child, container) {
	    const index = container.children.indexOf(child);
	    if (index === -1) {
	        throw new Error('child不存在');
	    }
	    container.children.splice(index, 1);
	}
	function insertChildToContainer(child, container, before) {
	    const beforeIndex = container.children.indexOf(before);
	    if (beforeIndex === -1) {
	        throw new Error('before不存在');
	    }
	    const index = container.children.indexOf(child);
	    if (index !== -1) {
	        container.children.splice(index, 1);
	    }
	    container.children.splice(beforeIndex, 0, child);
	}
	const scheduleMicroTask = typeof queueMicrotask === 'function'
	    ? queueMicrotask
	    : typeof Promise === 'function'
	        ? (callback) => Promise.resolve(null).then(callback)
	        : setTimeout;

	const supportSymbol = typeof Symbol === 'function' && Symbol.for;
	const REACT_ELEMENT_TYPE = supportSymbol
	    ? Symbol.for('react.element')
	    : 0xeac7;
	const REACT_FRAGMENT_TYPE = supportSymbol
	    ? Symbol.for('react.fragment')
	    : 0xeacb;

	function ChildReconciler(shouldTrackEffects) {
	    function deleteChild(returnFiber, childToDelete) {
	        if (!shouldTrackEffects) {
	            return;
	        }
	        const deletions = returnFiber.deletions;
	        if (deletions === null) {
	            returnFiber.deletions = [childToDelete];
	            returnFiber.flags |= ChildDeletion;
	        }
	        else {
	            deletions.push(childToDelete);
	        }
	    }
	    function deleteRemainingChildren(returnFiber, currentFirstChild) {
	        if (!shouldTrackEffects) {
	            return;
	        }
	        let childToDelete = currentFirstChild;
	        while (childToDelete !== null) {
	            deleteChild(returnFiber, childToDelete);
	            childToDelete = childToDelete.sibling;
	        }
	    }
	    function reconcileSingleElement(returnFiber, currentFiber, element) {
	        const key = element.key;
	        while (currentFiber !== null) {
	            if (currentFiber.key === key) {
	                if (element.$$typeof === REACT_ELEMENT_TYPE) {
	                    if (currentFiber.type === element.type) {
	                        let props = element.props;
	                        if (element.type === REACT_FRAGMENT_TYPE) {
	                            props = element.props.children;
	                        }
	                        const existing = useFiber(currentFiber, props);
	                        existing.return = returnFiber;
	                        deleteRemainingChildren(returnFiber, currentFiber.sibling);
	                        return existing;
	                    }
	                    deleteRemainingChildren(returnFiber, currentFiber);
	                    break;
	                }
	                else {
	                    {
	                        console.warn('未实现的react类型', element);
	                        break;
	                    }
	                }
	            }
	            else {
	                deleteChild(returnFiber, currentFiber);
	                currentFiber = currentFiber.sibling;
	            }
	        }
	        let fiber;
	        if (element.type === REACT_FRAGMENT_TYPE) {
	            fiber = createFiberFromFragment(element.props.children, key);
	        }
	        else {
	            fiber = createFiberFromElement(element);
	        }
	        fiber.return = returnFiber;
	        return fiber;
	    }
	    function useFiber(fiber, pendingProps) {
	        const clone = createWorkingProgress(fiber, pendingProps);
	        clone.index = 0;
	        clone.sibling = null;
	        return clone;
	    }
	    function reconcileSingleTextNode(returnFiber, currentFiber, content) {
	        while (currentFiber !== null) {
	            if (currentFiber.tag === HostText) {
	                const existing = useFiber(currentFiber, { content });
	                existing.return = returnFiber;
	                deleteRemainingChildren(returnFiber, currentFiber.sibling);
	                return existing;
	            }
	            deleteChild(returnFiber, currentFiber);
	            currentFiber = currentFiber.sibling;
	        }
	        const fiber = new FiberNode(HostText, { content }, null);
	        fiber.return = returnFiber;
	        return fiber;
	    }
	    function placeSingleChild(fiber) {
	        if (shouldTrackEffects && fiber.alternate === null) {
	            fiber.flags |= Placement;
	        }
	        return fiber;
	    }
	    function reconcileChildrenArray(returnFiber, currentFirstChild, 
	    // eslint-disable-next-line @typescript-eslint/no-explicit-any
	    newChild) {
	        let lastPlacedIndex = 0;
	        let lastNewFiber = null;
	        let firstNewFiber = null;
	        const existingChildren = new Map();
	        let current = currentFirstChild;
	        while (current !== null) {
	            const keyToUse = current.key !== null ? current.key : current.index;
	            existingChildren.set(keyToUse, current);
	            current = current.sibling;
	        }
	        for (let i = 0; i < newChild.length; i++) {
	            const after = newChild[i];
	            const newFiber = updateFromMap(returnFiber, existingChildren, i, after);
	            if (newFiber === null) {
	                continue;
	            }
	            newFiber.index = i;
	            newFiber.return = returnFiber;
	            if (lastNewFiber === null) {
	                lastNewFiber = newFiber;
	                firstNewFiber = newFiber;
	            }
	            else {
	                lastNewFiber.sibling = newFiber;
	                lastNewFiber = lastNewFiber.sibling;
	            }
	            if (!shouldTrackEffects) {
	                continue;
	            }
	            const current = newFiber.alternate;
	            if (current !== null) {
	                const oldIndex = current.index;
	                if (oldIndex < lastPlacedIndex) {
	                    newFiber.flags |= Placement;
	                    continue;
	                }
	                else {
	                    lastPlacedIndex = oldIndex;
	                }
	            }
	            else {
	                newFiber.flags |= Placement;
	            }
	        }
	        existingChildren.forEach((fiber) => {
	            deleteChild(returnFiber, fiber);
	        });
	        return firstNewFiber;
	    }
	    function updateFromMap(returnFiber, existingChildren, index, 
	    // eslint-disable-next-line @typescript-eslint/no-explicit-any
	    element) {
	        const keyToUse = element.key !== null ? element.key : index;
	        const before = existingChildren.get(keyToUse);
	        if (typeof element === 'string' || typeof element === 'number') {
	            if (before) {
	                if (before.tag === HostText) {
	                    existingChildren.delete(keyToUse);
	                    return useFiber(before, { content: element + '' });
	                }
	            }
	            return new FiberNode(HostText, { content: element + '' }, null);
	        }
	        if (typeof element === 'object' && element !== null) {
	            switch (element.$$typeof) {
	                case REACT_ELEMENT_TYPE:
	                    if (element.type === REACT_FRAGMENT_TYPE) {
	                        return updateFragment(returnFiber, before, element, keyToUse, existingChildren);
	                    }
	                    if (before) {
	                        if (before.type === element.type) {
	                            existingChildren.delete(keyToUse);
	                            return useFiber(before, element.props);
	                        }
	                    }
	                    return createFiberFromElement(element);
	            }
	        }
	        if (Array.isArray(element)) {
	            return updateFragment(returnFiber, before, element, keyToUse, existingChildren);
	        }
	        return null;
	    }
	    function updateFragment(returnFiber, current, 
	    // eslint-disable-next-line @typescript-eslint/no-explicit-any
	    element, key, existingChildren) {
	        let fiber;
	        if (!current || current.tag !== Fragment) {
	            fiber = createFiberFromFragment(element, key);
	        }
	        else {
	            existingChildren.delete(key);
	            fiber = useFiber(current, element);
	        }
	        fiber.return = returnFiber;
	        return fiber;
	    }
	    return function reconcilerChildFibers(returnFiber, currentFiber, 
	    // eslint-disable-next-line @typescript-eslint/no-explicit-any
	    newChild) {
	        const isUnkeyedTopLevelFragment = typeof newChild === 'object' &&
	            newChild !== null &&
	            newChild.type === REACT_FRAGMENT_TYPE &&
	            newChild.key === null;
	        if (isUnkeyedTopLevelFragment) {
	            newChild = newChild.props.children;
	        }
	        if (typeof newChild === 'object' && newChild !== null) {
	            if (Array.isArray(newChild)) {
	                return reconcileChildrenArray(returnFiber, currentFiber, newChild);
	            }
	            switch (newChild.$$typeof) {
	                case REACT_ELEMENT_TYPE:
	                    return placeSingleChild(reconcileSingleElement(returnFiber, currentFiber, newChild));
	                default:
	                    {
	                        console.warn('未实现的reconcile类型', newChild);
	                    }
	                    break;
	            }
	        }
	        if (typeof newChild === 'string' || typeof newChild === 'number') {
	            return placeSingleChild(reconcileSingleTextNode(returnFiber, currentFiber, newChild));
	        }
	        if (currentFiber !== null) {
	            deleteRemainingChildren(returnFiber, currentFiber);
	        }
	        {
	            console.warn('未实现的reconcile类型', newChild);
	        }
	        return null;
	    };
	}
	const reconcileChildFibers = ChildReconciler(true);
	const mountChildFibers = ChildReconciler(false);

	/* eslint-disable @typescript-eslint/no-explicit-any */
	const internals = React__namespace.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;

	const Passive = 0b0010;
	const HookHasEffect = 0b0001;

	/* eslint-disable @typescript-eslint/no-explicit-any */
	let currentlyRenderingFiber = null;
	let workInProgressHook = null;
	let currentHook = null;
	let renderLane = NoLane;
	const { currentDispatcher } = internals;
	function renderWithHooks(wip, lane) {
	    currentlyRenderingFiber = wip;
	    wip.memorizedState = null;
	    wip.updateQueue = null;
	    renderLane = lane;
	    const current = wip.alternate;
	    if (current !== null) {
	        currentDispatcher.current = HookDispatcherOnUpdate;
	    }
	    else {
	        currentDispatcher.current = HookDispatcherOnMount;
	    }
	    const Component = wip.type;
	    const props = wip.pendingProps;
	    const children = Component(props);
	    currentlyRenderingFiber = null;
	    workInProgressHook = null;
	    currentHook = null;
	    renderLane = NoLane;
	    return children;
	}
	const HookDispatcherOnMount = {
	    useState: mountState,
	    useEffect: mountEffect
	};
	const HookDispatcherOnUpdate = {
	    useState: updateState,
	    useEffect: updateEffect
	};
	function mountEffect(create, deps) {
	    const hook = mountWorkInProgressHook();
	    const nextDeps = deps === undefined ? null : deps;
	    currentlyRenderingFiber.flags |= PassiveEffect;
	    hook.memorizedState = pushEffect(Passive | HookHasEffect, create, undefined, nextDeps);
	}
	function updateEffect(create, deps) {
	    const hook = updateWorkInProgressHook();
	    const nextDeps = deps === undefined ? null : deps;
	    let destroy;
	    if (currentHook !== null) {
	        const prevEffect = currentHook.memorizedState;
	        destroy = prevEffect.destroy;
	        if (nextDeps !== null) {
	            const prevDeps = prevEffect.deps;
	            if (areHookInputsEqual(nextDeps, prevDeps)) {
	                hook.memorizedState = pushEffect(Passive, create, destroy, nextDeps);
	                return;
	            }
	        }
	        currentlyRenderingFiber.flags |= PassiveEffect;
	        hook.memorizedState = pushEffect(Passive | HookHasEffect, create, destroy, nextDeps);
	    }
	}
	function areHookInputsEqual(nextDeps, prevDeps) {
	    if (prevDeps === null || nextDeps === null) {
	        return false;
	    }
	    for (let i = 0; i < prevDeps.length && i < nextDeps.length; i++) {
	        if (Object.is(prevDeps[i], nextDeps[i])) {
	            continue;
	        }
	        return false;
	    }
	    return true;
	}
	function pushEffect(hookFlags, create, destroy, deps) {
	    const effect = {
	        tag: hookFlags,
	        create,
	        destroy,
	        deps,
	        next: null
	    };
	    const fiber = currentlyRenderingFiber;
	    const updateQueue = fiber.updateQueue;
	    if (updateQueue === null) {
	        const updateQueue = createFCUpdateQueue();
	        fiber.updateQueue = updateQueue;
	        effect.next = effect;
	        updateQueue.lastEffect = effect;
	    }
	    else {
	        const lastEffect = updateQueue.lastEffect;
	        if (lastEffect === null) {
	            effect.next = effect;
	            updateQueue.lastEffect = effect;
	        }
	        else {
	            const firstEffect = lastEffect.next;
	            lastEffect.next = effect;
	            effect.next = firstEffect;
	            updateQueue.lastEffect = effect;
	        }
	    }
	    return effect;
	}
	function createFCUpdateQueue() {
	    const updateQueue = createUpdateQueue();
	    updateQueue.lastEffect = null;
	    return updateQueue;
	}
	function updateState() {
	    const hook = updateWorkInProgressHook();
	    const queue = hook.updateQueue;
	    const pending = queue.shared.pending;
	    queue.shared.pending = null;
	    if (pending !== null) {
	        const { memorizedState } = processUpdateQueue(hook.memorizedState, pending, renderLane);
	        hook.memorizedState = memorizedState;
	    }
	    return [hook.memorizedState, queue.dispatch];
	}
	function updateWorkInProgressHook() {
	    let nextCurrentHook;
	    if (currentHook === null) {
	        const current = currentlyRenderingFiber?.alternate;
	        if (current !== null) {
	            nextCurrentHook = current?.memorizedState;
	        }
	        else {
	            nextCurrentHook = null;
	        }
	    }
	    else {
	        nextCurrentHook = currentHook.next;
	    }
	    if (nextCurrentHook === null) {
	        throw new Error(`组件${currentlyRenderingFiber?.type}本次执行时的Hook比上次执行时多`);
	    }
	    currentHook = nextCurrentHook;
	    const newHook = {
	        memorizedState: currentHook.memorizedState,
	        updateQueue: currentHook.updateQueue,
	        next: null
	    };
	    if (workInProgressHook === null) {
	        if (currentlyRenderingFiber === null) {
	            throw new Error('请在函数组件内调用hook');
	        }
	        else {
	            workInProgressHook = newHook;
	            currentlyRenderingFiber.memorizedState = workInProgressHook;
	        }
	    }
	    else {
	        workInProgressHook.next = newHook;
	        workInProgressHook = newHook;
	    }
	    return workInProgressHook;
	}
	function mountState(initialState) {
	    const hook = mountWorkInProgressHook();
	    let memorizedState;
	    if (initialState instanceof Function) {
	        memorizedState = initialState();
	    }
	    else {
	        memorizedState = initialState;
	    }
	    const queue = createUpdateQueue();
	    hook.updateQueue = queue;
	    hook.memorizedState = memorizedState;
	    const dispatch = dispatchSetState.bind(null, currentlyRenderingFiber, queue);
	    queue.dispatch = dispatch;
	    return [memorizedState, dispatch];
	}
	function dispatchSetState(fiber, updateQueue, action) {
	    const lane = requestUpdateLane();
	    const update = createUpdate(action, lane);
	    enQueueUpdate(updateQueue, update);
	    scheduledUpdateOnFiber(fiber, lane);
	}
	function mountWorkInProgressHook() {
	    const hook = {
	        memorizedState: null,
	        updateQueue: null,
	        next: null
	    };
	    if (workInProgressHook === null) {
	        if (currentlyRenderingFiber === null) {
	            throw new Error('请在函数组件内调用hook');
	        }
	        else {
	            workInProgressHook = hook;
	            currentlyRenderingFiber.memorizedState = workInProgressHook;
	        }
	    }
	    else {
	        workInProgressHook.next = hook;
	        workInProgressHook = hook;
	    }
	    return workInProgressHook;
	}

	const beginWork = (wip, renderLane) => {
	    switch (wip.tag) {
	        case HostRoot:
	            return updateHostRoot(wip, renderLane);
	        case HostComponent:
	            return updateHostComponent(wip);
	        case HostText:
	            return null;
	        case FunctionComponent:
	            return updateFunctionComponent(wip, renderLane);
	        case Fragment:
	            return updateFragment(wip);
	        default:
	            {
	                console.warn('beginWork未实现的类型');
	            }
	            break;
	    }
	};
	function updateFragment(wip) {
	    const nextChildren = wip.pendingProps;
	    reconcileChildren(wip, nextChildren);
	    return wip.child;
	}
	function updateHostRoot(wip, renderLane) {
	    const baseState = wip.memorizedState;
	    const updateQueue = wip.updateQueue;
	    const pending = updateQueue.shared.pending;
	    updateQueue.shared.pending = null;
	    const { memorizedState } = processUpdateQueue(baseState, pending, renderLane);
	    wip.memorizedState = memorizedState;
	    const nextChildren = wip.memorizedState;
	    reconcileChildren(wip, nextChildren);
	    return wip.child;
	}
	function updateHostComponent(wip) {
	    const nextProps = wip.pendingProps;
	    const nextChildren = nextProps.children;
	    reconcileChildren(wip, nextChildren);
	    return wip.child;
	}
	function updateFunctionComponent(wip, renderLane) {
	    const nextChildren = renderWithHooks(wip, renderLane);
	    reconcileChildren(wip, nextChildren);
	    return wip.child;
	}
	function reconcileChildren(wip, children) {
	    const current = wip.alternate;
	    if (current !== null) {
	        wip.child = reconcileChildFibers(wip, current?.child, children);
	    }
	    else {
	        wip.child = mountChildFibers(wip, null, children);
	    }
	}

	/* eslint-disable @typescript-eslint/no-explicit-any */
	let nextEffect = null;
	const commitMutationEffects = (finishedWork, root) => {
	    nextEffect = finishedWork;
	    while (nextEffect !== null) {
	        const child = nextEffect.child;
	        if ((nextEffect.subTreeFlags & (MutationMask | PassiveMask)) !== NoFlags &&
	            child !== null) {
	            nextEffect = child;
	        }
	        else {
	            up: while (nextEffect !== null) {
	                commitMutationEffectOnFiber(nextEffect, root);
	                const sibling = nextEffect.sibling;
	                if (sibling !== null) {
	                    nextEffect = sibling;
	                    break up;
	                }
	                nextEffect = nextEffect.return;
	            }
	        }
	    }
	};
	const commitMutationEffectOnFiber = (finishedWork, root) => {
	    const flags = finishedWork.flags;
	    if ((flags & Placement) !== NoFlags) {
	        commitPlacement(finishedWork);
	        finishedWork.flags &= ~Placement;
	    }
	    if ((flags & Update) !== NoFlags) {
	        commitUpdate(finishedWork);
	        finishedWork.flags &= ~Update;
	    }
	    if ((flags & ChildDeletion) !== NoFlags) {
	        const deletions = finishedWork.deletions;
	        if (deletions !== null) {
	            deletions.forEach((childToDelete) => {
	                commitDeletion(childToDelete, root);
	            });
	        }
	        commitUpdate(finishedWork);
	        finishedWork.flags &= ~ChildDeletion;
	    }
	    if ((flags & PassiveEffect) !== NoFlags) {
	        commitPassiveEffect(finishedWork, root, 'update');
	        finishedWork.flags &= ~PassiveEffect;
	    }
	};
	function commitPassiveEffect(fiber, root, type) {
	    if (fiber.tag !== FunctionComponent ||
	        (type === 'update' && (fiber.flags & PassiveEffect) === NoFlags)) {
	        return;
	    }
	    const updateQueue = fiber.updateQueue;
	    if (updateQueue !== null) {
	        if (updateQueue.lastEffect === null && true) {
	            console.error('当FC存在PassiveEffect flag时，不应该不存在effect');
	        }
	        root.pendingPassiveEffects[type].push(updateQueue.lastEffect);
	    }
	}
	function commitHookEffectList(flags, lastEffect, callback) {
	    let effect = lastEffect.next;
	    do {
	        if ((effect.tag & flags) === flags) {
	            callback(effect);
	        }
	        effect = effect.next;
	    } while (effect !== lastEffect.next);
	}
	function coomitHookEffectListUnmount(flags, lastEffect) {
	    commitHookEffectList(flags, lastEffect, (effect) => {
	        const destroy = effect.destroy;
	        if (typeof destroy === 'function') {
	            destroy();
	        }
	        effect.tag &= ~HookHasEffect;
	    });
	}
	function coomitHookEffectListDestroy(flags, lastEffect) {
	    commitHookEffectList(flags, lastEffect, (effect) => {
	        const destroy = effect.destroy;
	        if (typeof destroy === 'function') {
	            destroy();
	        }
	    });
	}
	function coomitHookEffectListCreate(flags, lastEffect) {
	    commitHookEffectList(flags, lastEffect, (effect) => {
	        const create = effect.create;
	        if (typeof create === 'function') {
	            effect.destroy = create();
	        }
	    });
	}
	function recordHostChildrenToDelete(childrenToDelete, unmountFiber) {
	    const lastOne = childrenToDelete[childrenToDelete.length - 1];
	    if (!lastOne) {
	        childrenToDelete.push(unmountFiber);
	    }
	    else {
	        let node = lastOne.sibling;
	        while (node !== null) {
	            if (unmountFiber === node) {
	                childrenToDelete.push(unmountFiber);
	            }
	            node = node.sibling;
	        }
	    }
	}
	function commitDeletion(childToDelete, root) {
	    const rootChildrenToDelete = [];
	    commitNestedComponent(childToDelete, (unmountFiber) => {
	        switch (unmountFiber.tag) {
	            case HostComponent:
	                recordHostChildrenToDelete(rootChildrenToDelete, unmountFiber);
	                return;
	            case HostText:
	                recordHostChildrenToDelete(rootChildrenToDelete, unmountFiber);
	                return;
	            case FunctionComponent:
	                commitPassiveEffect(unmountFiber, root, 'unmount');
	                return;
	            default:
	                {
	                    console.warn('未处理的unmount类型', unmountFiber);
	                }
	                break;
	        }
	    });
	    if (rootChildrenToDelete.length) {
	        const hostParent = getHostParent(childToDelete);
	        if (hostParent !== null) {
	            rootChildrenToDelete.forEach((node) => {
	                removeChild(node.stateNode, hostParent);
	            });
	        }
	    }
	    childToDelete.return = null;
	    childToDelete.child = null;
	}
	function commitNestedComponent(root, onCommitUnmount) {
	    let node = root;
	    // eslint-disable-next-line no-constant-condition
	    while (true) {
	        onCommitUnmount(node);
	        if (node.child !== null) {
	            node.child.return = node;
	            node = node.child;
	            continue;
	        }
	        if (node === root) {
	            return;
	        }
	        while (node.sibling === null) {
	            if (node.return === null || node.return === root) {
	                return;
	            }
	            node = node.return;
	        }
	        node.sibling.return = node.return;
	        node = node.sibling;
	    }
	}
	const commitPlacement = (finishedWork) => {
	    {
	        console.warn('执行placement操作', finishedWork);
	    }
	    function getHostSibling(fiber) {
	        let node = fiber;
	        // eslint-disable-next-line no-constant-condition
	        findSibling: while (true) {
	            while (node.sibling === null) {
	                const parent = node.return;
	                if (parent === null ||
	                    parent.tag === HostComponent ||
	                    parent.tag === HostRoot) {
	                    return null;
	                }
	                node = parent;
	            }
	            node.sibling.return = node.return;
	            node = node.sibling;
	            while (node.tag !== HostText && node.tag !== HostComponent) {
	                if ((node.flags & Placement) !== NoFlags) {
	                    continue findSibling;
	                }
	                if (node.child === null) {
	                    continue findSibling;
	                }
	                else {
	                    node.child.return = node;
	                    node = node.child;
	                }
	            }
	            if ((node.flags & Placement) === NoFlags) {
	                return node.stateNode;
	            }
	        }
	    }
	    const hostParent = getHostParent(finishedWork);
	    const sibling = getHostSibling(finishedWork);
	    if (hostParent !== null) {
	        insertOrAppendPlacementNodeIntoContainer(finishedWork, hostParent, sibling);
	    }
	};
	function getHostParent(fiber) {
	    let parent = fiber.return;
	    while (parent) {
	        const parentTag = parent.tag;
	        if (parentTag === HostComponent) {
	            return parent.stateNode;
	        }
	        if (parentTag === HostRoot) {
	            return parent.stateNode.container;
	        }
	        parent = parent.return;
	    }
	    {
	        console.warn('未找到host parent');
	    }
	    return null;
	}
	function insertOrAppendPlacementNodeIntoContainer(finishedWork, hostParent, before) {
	    if (finishedWork.tag === HostComponent || finishedWork.tag === HostText) {
	        if (before) {
	            insertChildToContainer(finishedWork.stateNode, hostParent, before);
	        }
	        else {
	            appendChildToContainer(hostParent, finishedWork.stateNode);
	        }
	        return;
	    }
	    const child = finishedWork.child;
	    if (child !== null) {
	        insertOrAppendPlacementNodeIntoContainer(child, hostParent);
	        let sibling = child.sibling;
	        while (sibling !== null) {
	            insertOrAppendPlacementNodeIntoContainer(sibling, hostParent);
	            sibling = sibling.sibling;
	        }
	    }
	}

	function markUpdate(fiber) {
	    fiber.flags |= Update;
	}
	const completeWork = (wip) => {
	    const newProps = wip.pendingProps;
	    const current = wip.alternate;
	    switch (wip.tag) {
	        case HostComponent:
	            if (current !== null && wip.stateNode) {
	                markUpdate(wip);
	            }
	            else {
	                const instance = createInstance(wip.type, newProps);
	                appendAllChildren(instance, wip);
	                wip.stateNode = instance;
	            }
	            bubbleProperties(wip);
	            return null;
	        case HostText:
	            if (current !== null && wip.stateNode) {
	                const oldText = current.memorizedProps.content;
	                const newText = newProps.content;
	                if (oldText !== newText) {
	                    markUpdate(wip);
	                }
	            }
	            else {
	                const instance = createTextInstance(wip.memorizedProps.content);
	                wip.stateNode = instance;
	            }
	            bubbleProperties(wip);
	            return null;
	        case HostRoot:
	        case FunctionComponent:
	        case Fragment:
	            bubbleProperties(wip);
	            return null;
	        default:
	            {
	                console.warn('未处理的completeWork情况', wip);
	            }
	            break;
	    }
	};
	function appendAllChildren(parent, wip) {
	    let node = wip.child;
	    while (node !== null) {
	        if (node?.tag === HostComponent || node?.tag === HostText) {
	            appendInitialChild(parent, node.stateNode);
	        }
	        else if (node.child !== null) {
	            node.child.return = node;
	            node = node.child;
	            continue;
	        }
	        if (node === wip) {
	            return;
	        }
	        while (node.sibling === null) {
	            if (node.return === null || node.return === wip) {
	                return;
	            }
	            node = node.return;
	        }
	        node.sibling.return = node.return;
	        node = node.sibling;
	    }
	}
	function bubbleProperties(wip) {
	    let subTreeFlags = NoFlags;
	    let child = wip.child;
	    while (child !== null) {
	        subTreeFlags |= child.subTreeFlags;
	        subTreeFlags |= child.flags;
	        child.return = wip;
	        child = child.sibling;
	    }
	    wip.subTreeFlags |= subTreeFlags;
	}

	/* eslint-disable @typescript-eslint/no-explicit-any */
	let syncQueue = null;
	let isFlushingSyncQueue = false;
	function scheduleSyncCallback(callback) {
	    if (syncQueue === null) {
	        syncQueue = [callback];
	    }
	    else {
	        syncQueue.push(callback);
	    }
	}
	function flushSyncCallbacks() {
	    if (!isFlushingSyncQueue && syncQueue) {
	        isFlushingSyncQueue = true;
	        try {
	            syncQueue.forEach((callback) => callback());
	        }
	        catch (e) {
	            {
	                console.error('flushSyncCallbacks报错', e);
	            }
	        }
	        finally {
	            isFlushingSyncQueue = false;
	            syncQueue = null;
	        }
	    }
	}

	let workInProgress;
	let wipRootRenderLane = NoLane;
	let rootDoesHasPassiveEffects = false;
	function prepareFreshStack(root, lane) {
	    workInProgress = createWorkingProgress(root.current, {});
	    wipRootRenderLane = lane;
	}
	function performSyncWorkOnRoot(root, lane) {
	    const nextLane = getHighestPriorityLane(root.pendingLanes);
	    if (nextLane !== SyncLane) {
	        ensureRootIsScheduled(root);
	        return;
	    }
	    {
	        console.warn('render阶段开始');
	    }
	    prepareFreshStack(root, lane);
	    do {
	        try {
	            workLoop();
	            break;
	        }
	        catch (e) {
	            {
	                console.warn('workloop发生错误', e);
	            }
	            workInProgress = null;
	        }
	        // eslint-disable-next-line no-constant-condition
	    } while (true);
	    const finishedWork = root.current.alternate;
	    root.finishedWork = finishedWork;
	    root.finishedLane = lane;
	    wipRootRenderLane = NoLane;
	    commitRoot(root);
	}
	function commitRoot(root) {
	    const finishedWork = root.finishedWork;
	    if (finishedWork === null) {
	        return;
	    }
	    {
	        console.warn('commit阶段开始', finishedWork);
	    }
	    const lane = root.finishedLane;
	    if (lane === NoLane && true) {
	        console.error('commit阶段finishedLane不应该是NoLane');
	    }
	    root.finishedWork = null;
	    root.finishedLane = NoLane;
	    markRootFinished(root, lane);
	    if ((finishedWork.flags & PassiveMask) !== NoFlags ||
	        (finishedWork.subTreeFlags & PassiveMask) !== NoFlags) {
	        if (!rootDoesHasPassiveEffects) {
	            rootDoesHasPassiveEffects = true;
	            Scheduler.unstable_scheduleCallback(Scheduler.unstable_NormalPriority, () => {
	                flushPassiveEffect(root.pendingPassiveEffects);
	                return;
	            });
	        }
	    }
	    const subTreeHasEffect = (finishedWork.subTreeFlags & MutationMask) !== NoFlags;
	    const rootHasEffect = (finishedWork.flags & MutationMask) !== NoFlags;
	    if (subTreeHasEffect || rootHasEffect) {
	        root.current = finishedWork;
	        commitMutationEffects(finishedWork, root);
	    }
	    else {
	        root.current = finishedWork;
	    }
	    rootDoesHasPassiveEffects = false;
	    ensureRootIsScheduled(root);
	}
	function flushPassiveEffect(pendingPassiveEffects) {
	    pendingPassiveEffects.unmount.forEach((effect) => {
	        coomitHookEffectListUnmount(Passive, effect);
	    });
	    pendingPassiveEffects.unmount = [];
	    pendingPassiveEffects.update.forEach((effect) => {
	        coomitHookEffectListDestroy(Passive | HookHasEffect, effect);
	    });
	    pendingPassiveEffects.update.forEach((effect) => {
	        coomitHookEffectListCreate(Passive | HookHasEffect, effect);
	    });
	    pendingPassiveEffects.update = [];
	}
	function workLoop() {
	    while (workInProgress !== null) {
	        performUnitOfWork(workInProgress);
	    }
	}
	function performUnitOfWork(fiber) {
	    const next = beginWork(fiber, wipRootRenderLane);
	    fiber.memorizedProps = fiber.pendingProps;
	    if (next === null) {
	        completeUnitOfWork(fiber);
	    }
	    else {
	        workInProgress = next;
	    }
	}
	function completeUnitOfWork(fiber) {
	    let node = fiber;
	    do {
	        completeWork(node);
	        const sibling = node.sibling;
	        if (sibling !== null) {
	            workInProgress = sibling;
	            return;
	        }
	        node = node.return;
	        workInProgress = node;
	    } while (node !== null);
	}
	function scheduledUpdateOnFiber(fiber, lane) {
	    const root = markUpdateFromFiberToRoot(fiber);
	    markRootUpdate(root, lane);
	    ensureRootIsScheduled(root);
	}
	function ensureRootIsScheduled(root) {
	    const updateLane = getHighestPriorityLane(root.pendingLanes);
	    if (updateLane === NoLane) {
	        return;
	    }
	    if (updateLane === SyncLane) {
	        {
	            console.log('在微任务中调度，优先级：', updateLane);
	        }
	        scheduleSyncCallback(performSyncWorkOnRoot.bind(null, root, updateLane));
	        scheduleMicroTask(flushSyncCallbacks);
	    }
	}
	function markRootUpdate(root, lane) {
	    root.pendingLanes = mergeLanes(root.pendingLanes, lane);
	}
	function markUpdateFromFiberToRoot(fiber) {
	    let node = fiber;
	    let parent = node.return;
	    while (parent !== null) {
	        node = parent;
	        parent = node.return;
	    }
	    if (node.tag === HostRoot) {
	        return node.stateNode;
	    }
	    return null;
	}

	function createContainer(container) {
	    const hostRootFiber = new FiberNode(HostRoot, {}, null);
	    const root = new FiberRootNode(container, hostRootFiber);
	    hostRootFiber.updateQueue = createUpdateQueue();
	    return root;
	}
	function updateContainer(element, root) {
	    const hostRootFiber = root.current;
	    const lane = requestUpdateLane();
	    const update = createUpdate(element, lane);
	    enQueueUpdate(hostRootFiber.updateQueue, update);
	    scheduledUpdateOnFiber(hostRootFiber, lane);
	    return element;
	}

	/* eslint-disable @typescript-eslint/no-explicit-any */
	let idCounter = 0;
	function createRoot() {
	    const container = {
	        rootID: idCounter++,
	        children: []
	    };
	    // eslint-disable-next-line @typescript-eslint/no-explicit-any
	    const root = createContainer(container);
	    function getChildren(parent) {
	        if (parent) {
	            return parent.children;
	        }
	        return null;
	    }
	    function getChildrenAsJSX(root) {
	        const children = childToJSX(getChildren(root));
	        if (Array.isArray(children)) {
	            return {
	                $$typeof: REACT_ELEMENT_TYPE,
	                type: REACT_FRAGMENT_TYPE,
	                key: null,
	                ref: null,
	                props: { children },
	                __mark: 'ly'
	            };
	        }
	        return children;
	    }
	    function childToJSX(child) {
	        if (typeof child === 'string' || typeof child === 'number') {
	            return child;
	        }
	        if (Array.isArray(child)) {
	            if (child.length === 0) {
	                return null;
	            }
	            if (child.length === 1) {
	                return childToJSX(child[0]);
	            }
	            const children = child.map(childToJSX);
	            if (children.every((child) => typeof child === 'string' || typeof child === 'number')) {
	                return children.join('');
	            }
	            return children;
	        }
	        if (Array.isArray(child.children)) {
	            const instance = child;
	            const children = childToJSX(instance.children);
	            const props = instance.props;
	            if (children !== null) {
	                props.children = children;
	            }
	            return {
	                $$typeof: REACT_ELEMENT_TYPE,
	                type: instance.type,
	                key: null,
	                ref: null,
	                props,
	                __mark: 'ly'
	            };
	        }
	        return child.text;
	    }
	    return {
	        _Scheduler: Scheduler__namespace,
	        render(element) {
	            return updateContainer(element, root);
	        },
	        getChildren() {
	            return getChildren(container);
	        },
	        getChildrenAsJSX() {
	            return getChildrenAsJSX(container);
	        }
	    };
	}

	var reactNoopRenderer = /*#__PURE__*/Object.freeze({
		__proto__: null,
		createRoot: createRoot
	});

	return reactNoopRenderer;

}));
